### PR TITLE
Added local_shapes as a new parameter to DistributedArray

### LINKED ISSUE #59: Add local_shape to DistributedArray
# Motivation
The class `DistributedArray` currently requires users to provide the global shape of the array (via `global_shape`), however it does not allow users to choose how such an array is split between ranks: `local_shape` is found by the `local_split` method to ensure best load balancing between ranks. There are however scenarios where users may want to choose `local_shape`, for example when they know they array will be re-interpreted as nd-array by some operator and the split would need to take into account the fact dimension zero should be split whilst the other dimensions should be on the same rank.

As such, I suggest to add `local_shape` as new input parameter for DistributedArray.

# Tasks

- [ ] Add `local_shape=None` as new input parameter for DistributedArray. If provided, force such splitting if not provided default to what we currently have.
- [ ] Add test and example in `plot_distributed_array.py`

### PR DESCRIPTION
**Pull request by rohanbabbar04**
Closes #59 
Implemented `local_shapes` parameter and changed all instances of DistributedArray to handle local_shapes
I see that `local_shapes` is a better option which will make the local_shape=local_shapes[rank] otherwise the default.
```
# Example
from pylops_mpi import DistributedArray

arr = DistributedArray(global_shape=(100, ), local_shapes=[(30, ), (40, ), (30, )])
print(arr.rank, arr)

# Output
0 <DistributedArray with global shape=(100,), local shape=(30,), dtype=<class 'numpy.float64'>, processes=[0, 1, 2])> 
1 <DistributedArray with global shape=(100,), local shape=(40,), dtype=<class 'numpy.float64'>, processes=[0, 1, 2])> 
2 <DistributedArray with global shape=(100,), local shape=(30,), dtype=<class 'numpy.float64'>, processes=[0, 1, 2])> 
```

In `DistributedArray.py`
- Added local_shapes as a List[Tuple]
- Added local_shapes to `to_dist` method
- Added a check for local_shapes so that they align with global shape
- Changed all instances to use local_shape
- Removed send/recv from `ravel` as now it is being handled by local_shapes

In `FirstDerivative.py`, `SecondDerivative.py` `cls_basic.py` and `plotting.py`
- Minor: Add local_shapes as parameter to DistributedArray

In `decorators.py`
- Updated the `reshaped` method by adding `forward` and `stacking` to handle reshaping/redistributing of stacking operators(I did it like that bcoz the code for them and derivatives was quite similar, if another decorator needs to be made to keep them separate then let me know)...
- Also made the code a little clean and easier to understand

In `VStack.py` and `BlockDiag.py`
- Now we give an option to handle it under the hood, also added local_shapes to y.
- Removed the Value Error as now that can be handled by the local_shapes and decorator.

Tests and example
- Updated tests for laplacian to check if it works for non-divisibility of size
- Added a test for local_shapes
- Added an example for local_shapes(since we are working with different ranks, adding raw numbers wasn't possible so just reversed the list of local_split shapes)

### CODE REVIEW

File: examples/plot_distributed_array.py [hunk 1]
```diff
@@ -39,6 +42,20 @@
...
 
+###############################################################################
+# You also have the option of directly including the ``local_shapes`` as a parameter
+# to the :py:class:`pylops_mpi.DistributedArray`. This will enable the assignment
+# of shapes to local arrays on each rank. However, it's essential to ensure that
+# the number of processes matches the length of ``local_shapes``, and that the
+# combined local shapes should align with the ``global_shape`` along the desired ``axis``.
+local_shape = local_split(global_shape, MPI.COMM_WORLD, Partition.SCATTER, 0)
+# Assigning local_shapes(List of tuples)
+local_shapes = MPI.COMM_WORLD.allgather(local_shape)[::-1]
```
Comment (mrava87) [inline examples/plot_distributed_array.py:53]: What is the reason for `[::-1]`?

File: tests/test_distributedarray.py [hunk 2]
```diff
@@ -120,6 +121,28 @@ def test_to_dist(par):
     assert dist_array.axis == par['axis']
 
 
+@pytest.mark.mpi(min_size=2)
+@pytest.mark.parametrize("par", [(par1), (par1j), (par2),
+                                 (par2j), (par3), (par3j)])
+def test_local_shapes(par):
+    """Test the `local_shapes` parameter in DistributedArray"""
+    loc_shapes = MPI.COMM_WORLD.allgather(local_split(par['global_shape'],
+                                                      MPI.COMM_WORLD, par['partition'], par['axis']))[::-1]
```
Comment (mrava87) [inline tests/test_distributedarray.py:130]: same question as before, why `[::-1]`

**Review by mrava87 — COMMENTED** [review]
Good stuff! 

This PR looks very good and I am happy to see how the `local_shapes` can help streamline various parts of the library.

I have left a few minor comments, and I have only one major one: what is the reasoning for passing to DistributedArray a list of tuples containing the local shapes for all ranks instead of passing only that of the current rank?

Basically, in other words, does it make sense to have the code pattern all the time user side:

```
local_shape = local_split(global_shape, MPI.COMM_WORLD, Partition.SCATTER, 0)
local_shapes = MPI.COMM_WORLD.allgather(local_shape)[::-1]
arr = pylops_mpi.DistributedArray(global_shape=global_shape, local_shapes=local_shapes, axis=0)
```

instead of 

```
local_shape = local_split(global_shape, MPI.COMM_WORLD, Partition.SCATTER, 0)
arr = pylops_mpi.DistributedArray(global_shape=global_shape, local_shape=local_shape, axis=0)
```

and have the gathering part inside the init method? 

I am sure you have probably thought about this, but at first sight I cannot find a reason why the latter would not work 🤔

**Comment by rohanbabbar04** [comment]
Hi @mrava87, Lets take an example of global_shape=100 with 3 processes and the person wants to split it in (30, ), (40,) and (30, )
Using `local_shapes`
```
arr = DistributedArray(global_shape=100, local_shapes=[(30,), (40,), (30,)])

```
When it comes to using the `local_shape` what I see(correct me if I am wrong)
```
if rank == 0:
   arr = DistributedArray(global_shape=100, local_shape=(30, )])
elif rank == 1:
   arr = DistributedArray(global_shape=100, local_shape=(40, )])
elif rank == 2:
   arr = DistributedArray(global_shape=100, local_shape=(30, )])
```
That is why I preferred the `local_shapes` parameter, plus I see that the users can write the local_shapes, thus also checking if all the local_shapes match the global shape

File: tests/test_distributedarray.py (hunk 2 above)
Comment (rohanbabbar04) [reply tests/test_distributedarray.py:130]: This was basically to check if our local_shapes feature is working fine(I don't think we can add raw numbers as num_processes will be different)..., I forgot to remove [::-1] as I was just testing..
so lets take an example after gathering local_split with global_shape = (10, 10) -> rank0 = (4, 10) -> rank1 = (3, 10) -> rank2 = (3, 10)
collecting all will give [(4, 10),  (3, 10), (3, 10)](which is default) I wanted to test for different local_shapes so I reversed the list giving [(3, 10), (3, 10), (4, 10)] which thus local_shape at each rank were switched and tried testing which worked fine...

**Comment by mrava87** [comment]
Oh good point. I was more thinking about the case a user would use local_split, but your point is valid and indeed sometimes users may want to choose custom splits themselves and the code with if statements is much less clean and elegant than providing all local shapes as you did. So nothing to be changed here :)

File: tests/test_distributedarray.py (hunk 2 above)
Comment (mrava87) [reply tests/test_distributedarray.py:130]: Great, I had assumed that was the case (which is a good test to do) but I just wanted to be sure… I would say either remove it or leave it as but just add a small comment

File: examples/plot_distributed_array.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mrava87) [reply examples/plot_distributed_array.py:53]: Same as for the test, but since this is an example I would remove it not to create any confusion

**Comment by rohanbabbar04** [comment]
> Oh good point. I was more thinking about the case a user would use local_split, but your point is valid and indeed sometimes users may want to choose custom splits themselves and the code with if statements is much less clean and elegant than providing all local shapes as you did. So nothing to be changed here :)

Good..
I will just make all the necessary changes which you asked for to make `local_shapes` work and commit soon..

File: tests/test_distributedarray.py (hunk 2 above)
Comment (rohanbabbar04) [reply tests/test_distributedarray.py:131]: Just added a small comment

File: pylops_mpi/basicoperators/BlockDiag.py [hunk 3]
```diff
@@ -110,23 +111,21 @@ def __init__(self, ops: Sequence[LinearOperator],
         dtype = _get_dtype(ops) if dtype is None else np.dtype(dtype)
         super().__init__(shape=shape, dtype=dtype, base_comm=base_comm)
 
+    @reshaped(forward=True, stacking=True)
     def _matvec(self, x: DistributedArray) -> DistributedArray:
-        if x.local_shape != (self.mops, ):
-            raise ValueError(f"Dimension mismatch: x shape-{x.local_shape} does not match operator shape "
-                             f"{self.localop_shape}; {x.local_shape[0]} != {self.mops} (dim1) at rank={self.rank}")
-        y = DistributedArray(global_shape=self.shape[0], dtype=x.dtype)
+        local_shapes = self.base_comm.allgather((self.nops, ))
```
[the remark this answers is not in the document]
Comment (rohanbabbar04) [reply pylops_mpi/basicoperators/BlockDiag.py]: Done in `BlockDiag` as well as `VStack.py`

File: pylops_mpi/DistributedArray.py [hunk 4]
```diff
@@ -274,6 +292,23 @@ def to_dist(cls, x: NDArray,
...
+    def _check_local_shapes(self, local_shapes):
+        if local_shapes:
+            if len(local_shapes) != self.base_comm.size:
+                raise ValueError(f"Length of local shapes is not equal to number of processes; "
+                                 f"{len(local_shapes)} != {self.size}")
+            # Check if local shape == global shape
+            if self.partition is Partition.BROADCAST and local_shapes[self.rank] != self.global_shape:
+                raise ValueError(f"Local shape is not equal to global shape at rank = {self.rank};"
+                                 f"{local_shapes[self.rank]} != {self.global_shape}")
+            if self.partition is Partition.SCATTER:
```
[the remark this answers is not in the document]
Comment (rohanbabbar04) [reply pylops_mpi/DistributedArray.py]: Done..

File: pylops_mpi/DistributedArray.py [hunk 5]
```diff
@@ -274,6 +292,23 @@ def to_dist(cls, x: NDArray,
             dist_array[:] = x[tuple(slices)]
         return dist_array
 
+    def _check_local_shapes(self, local_shapes):
```
[the remark this answers is not in the document]
Comment (rohanbabbar04) [reply pylops_mpi/DistributedArray.py:295]: Added a docstring

File: examples/plot_distributed_array.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (rohanbabbar04) [reply examples/plot_distributed_array.py]: Removed `[::-1]`

File: pylops_mpi/DistributedArray.py [hunk 6]
```diff
@@ -274,6 +292,23 @@ def to_dist(cls, x: NDArray,
...
+            # Check if local shape == global shape
+            if self.partition is Partition.BROADCAST and local_shapes[self.rank] != self.global_shape:
+                raise ValueError(f"Local shape is not equal to global shape at rank = {self.rank};"
+                                 f"{local_shapes[self.rank]} != {self.global_shape}")
+            if self.partition is Partition.SCATTER:
+                local_shape = local_shapes[self.rank]
+                # Check if local shape sum up to global shape and other dimensions align with global shape
+                if self.base_comm.allreduce(local_shape[self.axis]) != self.global_shape[self.axis] or \
+                        not np.array_equal(np.delete(local_shape, self.axis), np.delete(self.global_shape, self.axis)):
+                    raise ValueError(f"Local shapes don't sum up to global shape;"
```
[the remark this answers is not in the document]
Comment (rohanbabbar04) [reply pylops_mpi/DistributedArray.py]: Done changed 
```
raise ValueError(f"Local shapes don't align with the global shape;"
                                     f"{local_shapes} != {self.global_shape}")
```

**Review by mrava87 — APPROVED** [review]

**Pull request merged**